### PR TITLE
Add examples for built-in types

### LINKED ISSUE #9: Add examples for the built-in types mentioned in the readme
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/type-fest/issues/9)

I made a section in the readme for useful built-in types that many TS devs don't know about. Would be useful to also include some real-world examples of those types in use. It's hard to realize when they would be useful just from their description.

https://github.com/sindresorhus/type-fest#built-in-types

To resolve this issue you should at least add one realistic example for each type.

> There is a $60.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/type-fest/issues/9).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/type-fest/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by krnik**
Fixes #9

Added examples to built-in types.
Please, let me know if there are things to correct.

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -86,17 +86,495 @@ Click the type names for complete docs.
...
+    </summary>
+
+    ```typescript
+    interface Article {
+        title: string;
+        thumbnail: string;
+        content: string;
+    }
+
+    type ArticlePreview = Pick<Article, 'title' | 'thumbnail'>;
```
Comment (sindresorhus) [inline readme.md]: Would be good to have a comment there that says that this creates a new type with the chosen properties from `Article`.

File: readme.md [hunk 2]
```diff
@@ -86,17 +86,495 @@ Click the type names for complete docs.
...
+
 - [`Record<K, T>`](https://github.com/Microsoft/TypeScript/blob/2961bc3fc0ea1117d4e53bc8e97fa76119bc33e3/src/lib/es5.d.ts#L1429-L1434) - Construct a type with a set of properties `K` of type `T`.
+    <details>
+    <summary>
+        Example
+        <a href="https://typescript-play.js.org/?target=6#code/C4TwDgpgBAshC2AjCAnACgewM4Etg4wDssoBeKAIh0OFUIqgB9KATCANwgBsNIUHmFWgGMAFgFouEAIYsKAbgBQi6rRQAzacOhpUWIlADeiqFHU4UWYADlp8CAC4oVlNQDmS012lXb9py7unlAgMpYA8uoAogAefDgQhNpOhACuSKhKAL7KwkRWULR2TgBKEHkoLAA8cBno2HgExAA0ULqWRADaALoAfGRGJlCqdE49zUNsnDx8Y90Tpp1C5RJSshTdc4pZSooA9HtQ4akohTLwUKI+lxYQLFDSUIQQAO5Q9nUAdIpF8J8jKEInzAqSwogAFIYzBZfHZHJQAFIYUT0VreWH+SgAEQwEAorVC0gi0TiqASSXhAAYoFkAJS7A5QMoVe7qDAobQkEAYVKFDDDQiNaRcHAAL2gwq4UAw6kKomgYBQvFQoCgAGsICAsPtDgAVcAQADKwlcYGAUCiKCVKCcyzEkhkcigiuVKFVOBI8A9uEIbkUeWI5t+UXgZpApXK7OqtWQ9Vw+HyrTSXC4-XIxlMpgBhBSqRTC0zU24rtz+e28iAA">
+            Playground
+        </a>
+    </summary>
+
```
Comment (sindresorhus) [inline readme.md]: For each type it would be useful to include a non-technical sentence of what it does here. ``` Construct a type with a set of properties `K` of type `T` ``` describes the type in type-system terms, but not in human terms.

File: readme.md [hunk 3]
```diff
@@ -86,17 +86,495 @@ Click the type names for complete docs.
...
+    interface Person {
+        firstName: string;
+        lastName: string;
+        yearsOfExperience: number;
+    }
+
+    const team: Record<MemberPositions, Person[]> = {
+        intern: [],
+        developer: [],
+        ['tech-lead']: [],
```
Comment (sindresorhus) [inline readme.md]:
```suggestion
        'tech-lead': [],
```

File: readme.md [hunk 4]
```diff
@@ -86,17 +86,495 @@ Click the type names for complete docs.
...
+    type KeysOfType<ObjectType, TargetType> = {
+        [Key in keyof ObjectType]: ObjectType[Key] extends TargetType ? Key : never;
+    }[keyof ObjectType];
+
+
+    /** Omit those properties that extend ExcludeType */
+    type OmitTypeKeys<ObjectType, ExcludeType> = Pick<
+        ObjectType,
+        Exclude<keyof ObjectType, KeysOfType<ObjectType, ExcludeType>>
+    >;
```
Comment (sindresorhus) [inline readme.md]: I don't think the example should require helper types. This example is more complicated than it needs to be.

File: readme.md [hunk 5]
```diff
@@ -86,17 +86,495 @@ Click the type names for complete docs.
...
+    /** From keys of ObjecTypeA remove all keys that are present on ObjecTypeB */
+    type AdditionalKeys<ObjectTypeA, ObjectTypeB> = Exclude<
+        keyof ObjectTypeA,
+        keyof ObjectTypeB
+    >;
+
+    /** Keep only those keys of ObjectTypeA that are not present on ObjecTypeB */
+    type Diff<ObjectTypeA, ObjectTypeB> = {
+        [Key in AdditionalKeys<ObjectTypeA, ObjectTypeB>]: ObjectTypeA[Key]
+    };
```
Comment (sindresorhus) [inline readme.md]: This is not directly related to `Exclude`. I would prefer the example to be simple without this.

File: readme.md [hunk 6]
```diff
@@ -86,17 +86,495 @@ Click the type names for complete docs.
...
+    <details>
+    <summary>
+        Example
+        <a href="https://typescript-play.js.org/?target=6#code/PTAEHMBsHsCMENIBUAWBLAzqeA3ebJ5ZIBTUAMwCdoBbUAZgDoAWAKBFBQBcuAHDAFwgAJiRzFo4DIxpoAxtQzRyXRnNrAuAT14kMCtLy7B4AO1PQArqblpT4ALTbd+yoa4P6D5sADEzkgc5FBI5AGs7R3JoSgcoOEQudAxWdVMMLggYBGRk0ABeUAB3O2FoIoBuVnYAKhrQACFSims5LjRoU1AAA3JTbtAuaEGQ0DSuEgAPTOUshMhQOAArUMya4FZyVvbO0FhSpGgAcWzEAB4AMS6pidNhLAAKRmf4SikBbFMtAG0AXQBKAoAPk+WiBD1YoAopg+VwANJDQM9GK93qAAAqveA0EgTSgYS6mIGsf4fB6A-IggBKuMslFMSB0JEJIIA3ojKLT6dDGPs7g94jlUJg4UiXm8MP8qgBfarAOqgADCIXCoDQ5DGnQm0048CwvGoODQomEPTCJC03UYoHWm22HS6wVCYQAkuR0dReABRSaYLgYB5JTAfZarUXmrQfDJueykvbQaCkMygdlQzlcOldCNqrpBjAyuUKmkZ+lYbpcSiWEgDdVzHKLWArNq6ssAcnIuOCrYGBuguko2htGzSGTGKtd5AunZQPr9WEKfOEhxO8weTvCbo9fdnGQwooARB2uMF91LquuJ1PjzPfbvyRVQBwK1Wc6AaNBR7MkmRYNQihgSHxVggA">
+            Playground
+        </a>
+    </summary>
+
+    ```typescript
+    /** Bind function `fn` to the context of global object */
```
Comment (sindresorhus) [inline readme.md]: Could you come up with a more realistic example than this? I don't think anyone actually bind to a global like this.

File: readme.md [hunk 7]
```diff
@@ -86,17 +86,495 @@ Click the type names for complete docs.
...
 - [`ConstructorParameters<T>`](https://github.com/Microsoft/TypeScript/blob/2961bc3fc0ea1117d4e53bc8e97fa76119bc33e3/src/lib/es5.d.ts#L1456-L1459) - Obtain the parameters of a constructor function type in a tuple.
+    <details>
+    <summary>
+        Example
+        <a href="https://typescript-play.js.org/?target=6#code/MYGwhgzhAECCBOAXAlqApgWQPYBM0mgG8AoaaYLAO0TWoC5oJF5lKBzAblOhURDQZMW7LtwqUhAV2CIs8ABTia9Rs1ZsAND2R8Bq4WwCURbmUQALZBAB0S2omgBeclWWIuZM5Zu9+T7boe0AC+xKHE4kzQAA7wWAAeAJ4AEmCUOPzw-iRkkczSiPKm0AAqYPBsaIgMiInRaFgAZnBIqPzYeCAaxeVsEAwAwlRSMnIACuVgALZVaPAQADy19U0tKOgd+AB83Z7QlGgA7mUVVQxpidzGOZ6RWPzWIFhs8gAGAJKUOshgIMgAXupoAASQgIdbtXD4ayUaZoYLWV6GIJkeBVSTwSj7I6lXpVeTWQm9CDI7ihYKiAD0lNKdTQAGVgCxog4AKLwOLwBis+L1GRoHDQADaQnUWlF7AAulo2FgHCK1FLiLEEik0hk5rZhvkZPJwW1MFCusKAORDCQ6xCE6wm6VrA2bECk4hoXlyBx5e3ofwHQ7QMZxJJ61obI1aFVJVLpTLO31e-jyABEJTpjOZDgAzNYACzWACs1sTWkTABk0OUsSn6mnkCzoAA5LCHACEidJ1OgAHVkCACE82AwPl8UL8AUD9aHOojiEA">
+            Playground
+        </a>
+    </summary>
+
+    ```typescript
```
Comment (sindresorhus) [inline readme.md]: I would really prefer a simpler example than using Proxy.

**Comment by sindresorhus** [comment]
Nice work! :)

**Comment by sindresorhus** [comment]
Can you also include the new `Omit` type in TS 3.5?

**Comment by krnik** [comment]
Sure, I'll include `Omit` and try to come up with simpler and more clear examples.
I've never thought it was such a hard task to create great examples and description ^^

**Comment by krnik** [comment]
Okay, hope I didn't miss anything.

I moved `Playground` link below `Example` - it was actually unnecessary to show the link if someone didn't want to see the example.
Also, I changed some of the examples following your advice.

**Comment by sindresorhus** [comment]
Thanks for your awesome work on this :)
